### PR TITLE
Fix issue 15889 - Array bounds check should report index and length

### PR DESCRIPTION
**Pull request by dkorpel**
Uses the `ArrayIndexError` and `ArraySliceError` instead of `RangeError` for array index / slice operations.
Depends on this druntime PR: https://github.com/dlang/druntime/pull/3517

I still need to decide when to use them:
- always
- with `-checkaction=context`
- in debug builds
- in unoptimized builds
- ...?

### CODE REVIEW

**Comment by MoonlightSentinel** [comment]
> I still need to decide when to use them:

IMO the bounds should always be reported when a `RangeError` is thrown. They can provide vital information for a neglegible cost (in contrast to the whole `RangeError` construction + throwing).

**Comment by RazvanN7** [comment]
The vision document for druntime states that implementing druntime hooks as templates is a major objective. Would it make sense for these hooks to be implemented as templates? (the main advantage would be that the associated druntime functions will not be part of the executable if not used.)

**Comment by dkorpel** [comment]
> The vision document for druntime states that implementing druntime hooks as templates is a major objective. Would it make sense for these hooks to be implemented as templates?

I thought the idea was to replace functions taking TypeInfo with functions templated on the type. In this case, since range errors don't depend on the array type, a normal function works fine. The linker can still strip `_d_arraybounds_indexp` if it's not used (though I think very few D projects don't need bounds checks).

**Comment by dkorpel** [comment]
> /home/runner/work/dmd/dmd/dmd/test/test_results/runnable_cxx/cppa_0: symbol lookup error: /home/runner/work/dmd/dmd/dmd/test/test_results/runnable_cxx/cppa_0: undefined symbol: _d_arraybounds_slicep

Looks like the C++ interop tests with GCC aren't linking with the newest druntime, does anyone know how to fix it?

**Comment by dkorpel** [comment]
> Looks like the C++ interop tests with GCC aren't linking with the newest druntime, does anyone know how to fix it?

It turns out the checkout is fine, but it doesn't seem to be linking with druntime to begin with. I worked around it by adding `-checkaction=C` to the cppa.d test. Thanks to @Geod24 for his assistance!

File: changelog/range-error.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
+Out of bounds array access now gives a fluent error message
```
Comment (Geod24) [inline changelog/range-error.md:1]: POV: "fluent" is a bit vague, I would advise to either phrase it as "now includes indices" or mention "better" or "improved" error message. But that's POV, I think the message is good enough as-is.

File: changelog/range-error.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+$(CONSOLE
+> dmd -run main.d
+core.exception.ArraySliceError@main.d(4): slice [2 .. 4] extends past array of length 3
+$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)
+??:? _d_arraybounds_slicep [0x5647250b980d]
+??:? _Dmain [0x5647250b9751]
+)
+
+The error message for indexing a non-existent key in an Associative Array has not been updated.
```
Comment (Geod24) [inline changelog/range-error.md:54]: Not sure why this mention was warranted ?

File: changelog/range-error.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+Previously this would yield the following error when compiled and run:
+
+$(CONSOLE
+> dmd -run main.d
+core.exception.RangeError@main.d(4): Range violation
+$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)$(NDASH)
+??:? _d_arrayboundsp [0x555765c167f9]
+??:? _Dmain [0x555765c16752]
+)
```
Comment (Geod24) [inline changelog/range-error.md:25]: The changelog doesn't mention the two new classes derived from `Error`, except for the implicit mentions in the code sample. Not sure if you want to add it here or not.

**Review by Geod24 — APPROVED** [review]
Can you merge the second commit ("Add runnable test...") with the fix ? A fix / new feature should come with tests in the same commit. Also, could you swap the order of the commits so that the last commit comes first ? Each commit should be atomic, which means they should compile and pass the test-suite on their own, and the last commit is required for your feature to pass the test suite (but the last commit makes sense on its own, as the same problem would arise if, e.g. `-checkaction=context` was made the default, so 👍 for having it as a separate commit).

The code LGTM, and so does the changelog. Thanks for putting time to make it both informative and appealing, and for tackling that feature, it'll make a huge difference in usability.

File: changelog/range-error.md (hunk 2 above)
Comment (dkorpel) [reply changelog/range-error.md:54]: I figured people might wonder if all range errors are improved, so I'm stating a known limitation here.

File: changelog/range-error.md (hunk 1 above)
Comment (dkorpel) [reply changelog/range-error.md:1]: I borrowed the term 'fluent' from the common 'fluent asserts' you find in unit testing packages, but I agree it's not the best word. I'm thinking changing it to "informative" or, indeed, "better".

File: changelog/range-error.md (hunk 3 above)
Comment (dkorpel) [reply changelog/range-error.md:25]: I mentioned them at first, but I removed it because I don't want to encourage D users to catch range errors.

**Comment by dkorpel** [comment]
> Each commit should be atomic, which means they should compile and pass the test-suite on their own

Thanks for the advice, I don't know a good commit separation policy so I was just winging it. I will change it.

**Review by Geod24 — APPROVED** [review]

**Comment by ibuclaw** [comment]
This PR made RangeErrors with pointer slices more confusing, not less.
```
void main()
{
    int* a = [1,2,3,4,5,6,7,8,9,10].ptr;
    a[1..5] = a[3..7];
}
```
```
core.exception.ArrayIndexError@range.d(4): index [4] exceeds array of length 4
```

**Comment by dkorpel** [comment]
I'll look into it

File: src/dmd/e2ir.d [hunk 4]
```diff
@@ -2246,7 +2246,7 @@ extern (C++) class ToElemVisitor : Visitor
                         }
 
                         // Construct: (c || arrayBoundsError)
-                        echeck = el_bin(OPoror, TYvoid, c, buildArrayBoundsError(irs, ae.loc, null, el_copytree(eleny), el_copytree(elen)));
+                        echeck = el_bin(OPoror, TYvoid, c, buildArrayIndexError(irs, ae.loc, el_copytree(eleny), el_copytree(elen)));
```
Comment (dkorpel) [inline src/dmd/e2ir.d:2249]: This makes no sense, this is not an index error but a slice copy error.

**Pull request merged**